fabtests/common/shared.c

```diff
@@ -2983,9 +2994,31 @@ void ft_longopts_usage()
 	return;
 }
 
-void ft_parseinfo(int op, char *optarg, struct fi_info *hints,
+/* On Windows memory allocated outside the libfabric library must be freed outside
+ * the library. Lots of tests allow setting parts of the hints from command line
+ * arguments. This function will consolidate all memory allocations by calling
+ * fi_dupinfo, then free the parts that might have been allocated by the
+ * application so that it is safe to call fi_freeinfo.
+ */
+struct fi_info *ft_consolidate_hints(struct fi_info *hints, void **ptr)
+{
+	struct fi_info *consolidated_hints = fi_dupinfo(hints);
+
+	if (*ptr) {
+		free(*ptr);
+		*ptr = NULL;
+	}
+
+	fi_freeinfo(hints);
+
+	return consolidated_hints;
+}
```

Comment:
This flow is odd.  It would be easier to follow if the apps simply freed the memory that it attaches to the hints, versus calling libfabric to allocate and copy the structure, freeing it here, and returning the copy back.  Nothing in this call path suggests that this will happen, making this fragile.  If hints weren't allocated earlier using fi_allocinfo() or if the app stored a pointer to the original hints, this flow breaks.

Reply:
If only it were easy and straightforward to implement that.

Several places in shared.c will modify the global hints structure and the apps won't necessarily know about it, which makes it difficult for an app to know what to free before calling fi_freeinfo or ft_free_res, which will call fi_freeinfo for the app. Then there is the oddness of hints being a global, but then not treated that way by being used as a parameter to lots of functions. And all those functions are modifying hints as a side-effect despite the fact that hints is passed by value, so I think perhaps passing &hints actually makes it more clear that side effects are happening. Several apps tinker directly with hints on their own. And some call fi_dupinfo, which then makes all the memory of the returned fi_info pointer owned by libfabric and not the app.

Given all the various and sundry ways that hints was being used, it seemed like perhaps the most straightforward way of dealing with it was acknowledging that the complicating factor was the app adding its own pointer to part of the hints and making this an up front detail. If you have a hints fi_info pointer and you want to add an app owned pointer to it, add the pointer, then call ft_consolidate_hints with that pointer and it will return a new pointer that is safe to call fi_freeinfo with as well as freeing the app owned pointer.

> Nothing in this call path suggests that this will happen, making this fragile.

I would submit that a function taking a fi_info* and returning an fi_info* is pretty suggestive that perhaps an old for new swap is happening.
> If hints weren't allocated earlier using fi_allocinfo() ..., this flow breaks.

Does anyone really handcraft a fi_info hints structure and not just start with a call fi_allocinfo()? All the tests I saw had a line like `hints = fi_allocinfo()` near the beginning of main(). But your core point is definitely correct that if an app did create all its own bits of an fi_info struct, it wouldn't be correct to then pass that to fi_freeinfo.
> ...or if the app stored a pointer to the original hints, this flow breaks.

As I mentioned above, my idea was to sort of change the paradigm so that apps that needed to add app owned pieces to the hints would always add them then immediately swap that for a new structure that was safe to pass to fi_freeinfo and free the app owned data. So, under the new paradigm, it wouldn't be valid for an app to store the pointer to the original hints.

I hope none of that came off as being too defensive. I was trying hard to obtain the desired result and give windows what it needed with minimal code impact, which, I'll admit, isn't always the best way to good code. Sometimes I think that windows is the reason we can't have nice things. ;-)

I'm certainly open to any other ideas that you might have.

Reply:
I was thinking of something like this:
```
/* We need to free any data that we allocated before freeing the
 * hints.  Windows doesn't like it when a library frees memory that
 * was allocated by the application.
 */
void ft_freehints(struct fi_info *hints) 
{
    if (!hints)
        return;

    if (hints->src_addr) {
        free(hints->src_addr);
        hints->src_addr = NULL;
        hints->src_addrlen = 0;
    }
    if (hints->dest_addr) {
        ...
    }
    fi_freeinfo(hints);
}
```
Then update all locations where fi_freeinfo(hints) is called to call this instead.  The fi_allocinfo() call allocates the fi_info structure and all attribute substructures.  There are only a couple places where the hints would point to application allocated memory.  The src/dest addresses are two.  The auth_key might be the only other place.  Everything else is allocated by libfabric.

To dup hints, I would follow a similar flow.  I.e. create an ft_duphints(struct fi_info *hints) and have that perform the duplication.  This could be used in the shared_ctx test where we dup the hints to better isolate what's going on.

As a side note, fabtests does a poor job with this, but we could use const to indicate what parameters are not supposed to change.